middleware/recovery.go

```diff
@@ -0,0 +1,129 @@
+package middleware
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"runtime/debug"
+	"sync"
+
+	"github.com/danielgtaylor/huma"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+// MaxLogBodyBytes logs at most this many bytes of any request body during a
+// panic when using the recovery middleware. Defaults to 10KiB.
+var MaxLogBodyBytes int64 = 10 * 1024
+
+// bufferedReadCloser will read and buffer up to max bytes into buf. Additional
+// reads bypass the buffer.
+type bufferedReadCloser struct {
+	reader io.ReadCloser
+	buf    *bytes.Buffer
+	max    int64
+}
+
+// newBufferedReadCloser returns a new BufferedReadCloser that wraps reader
+// and reads up to max bytes into the buffer.
+func newBufferedReadCloser(reader io.ReadCloser, buffer *bytes.Buffer, max int64) *bufferedReadCloser {
+	return &bufferedReadCloser{
+		reader: reader,
+		buf:    buffer,
+		max:    max,
+	}
+}
+
+// Read data into p. Returns number of bytes read and an error, if any.
+func (r *bufferedReadCloser) Read(p []byte) (n int, err error) {
+	// Read from the underlying reader like normal.
+	n, err = r.reader.Read(p)
+
+	// If buffer isn't full, add to it.
+	length := int64(r.buf.Len())
+	if length < r.max {
+		if length+int64(n) < r.max {
+			r.buf.Write(p[:n])
+		} else {
+			r.buf.Write(p[:int64(n)-(r.max-length)])
+		}
+	}
+
+	return
+}
+
+// Close the underlying reader.
+func (r *bufferedReadCloser) Close() error {
+	return r.reader.Close()
+}
+
+// PanicFunc defines a function to run after a panic, which allows you to set
+// up custom logging, metrics, etc.
+type PanicFunc func(ctx context.Context, err error, request string)
+
+// Recovery prints stack traces on panic when used with the logging middleware.
+func Recovery(onPanic PanicFunc) func(http.Handler) http.Handler {
+	bufPool := sync.Pool{
+		New: func() interface{} {
+			return new(bytes.Buffer)
+		},
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var buf *bytes.Buffer
+
+			// Reset the body so other middleware/handlers can use it.
+			if r.Body != nil {
+				// Get a buffer that the body will be read into.
+				buf = bufPool.Get().(*bytes.Buffer)
+				defer bufPool.Put(buf)
+
+				r.Body = newBufferedReadCloser(r.Body, buf, MaxLogBodyBytes)
```

Comment:
If this is a panic, we should probably ensure we're logging the whole request even if it's bigger than 10k bytes.

Reply:
This is probably worth a larger discussion. When designing a web framework you have to balance speed / memory use / concurrent connection capacity / etc. Currently this uses a shared auto-scaling pool of byte buffers that read up to a certain amount into memory for every single incoming request, and that's not cheap. Given the size of most API requests 10KB was chosen as a sane default that should print the whole request most of the time, but prevents large bodies from taking down the service. This is definitely a compromise that attempts to give us at least some extremely useful debugging data at the expense of some speed and higher memory thrashing. Benchmarks show that we can process a couple thousand fewer requests a second because of it (a few percent).

I understand our current use-case isn't extreme high performance, but keep in mind this is meant to be a general-purpose web framework and could see adoption across many teams and companies.

Reply:
If the impact of changing this value is that big, this feels like the kind of switch that comes with a warning label.

I don't know if there's anything in the README about it, but a comment about the impact that changing this value could have on system performance might be worth including. If it's already there, apologies on my having missed it.